### ISSUE TITLE
/v1/config/proxy-defaults/global returns `json: unsupported type: map[interface {}]interface {}`

### ISSUE THREAD

**Issue by banks** [hashicorp/consul#5743]
Current master, despite all the efforts and tests in #5718 has a JSON decoding error.

## Reproduction Steps

 - From current master (51391b647)
 - update `version/version.go` to make it version 1.5.0 - failing to do this means the auto-pilot version check doesn't allow the bootstrap registration.
 - `make dev`
 - Run dev agent with following in config file:
```hcl
config_entries {
  bootstrap {
    proxy_defaults "global" {
      config {
        envoy_prometheus_bind_addr = "0.0.0.0:1234"
      }
    }
  }
}
```
 - 
```
curl -i localhost:8500/v1/config/proxy-defaults/global
HTTP/1.1 500 Internal Server Error
Vary: Accept-Encoding
Date: Tue, 30 Apr 2019 11:50:33 GMT
Content-Length: 53
Content-Type: text/plain; charset=utf-8

json: unsupported type: map[interface {}]interface {}
```

**Comment by banks** [comment]
This is because of the confusion around the nested `config`. The following config works OK:

```
config_entries {
  bootstrap {
    proxy_defaults "global" {
      envoy_prometheus_bind_addr = "0.0.0.0:1234"
    }
  }
}
```

This points to one bug being:
 - nested maps in central config break encoding JSON

In addition there is confusion around how to specify the proxy_defaults correctly. The model for the config entry is: 

```go
type ProxyConfigEntry struct {
	Kind   string
	Name   string
	Config map[string]interface{}
}
```

Where the only top-level key defined is the `Config` field but more may be added later. So the more verbose/ugly version I tried which reveals the bug is really the correct one - other config_entry types or this one in the future will have other fields that need to be set here.

This comes from here: https://github.com/hashicorp/consul/blob/master/agent%2Fconfig%2Fbuilder.go#L639-L649 the value from the config is used verbatim as the proxy config type's config field rather than allowing full specification of the whole config type.

Will think about the cleanest way to reconcile this.

**Comment by banks** [comment]
I think the best option is to change the config syntax to make the whole policy specified in the same way as you would via the CLI/API. I should have noticed this in review but a complete example (including future extensions for completeness would be):

```
config_entries {
  bootstrap {
    kind = "proxy-defaults"
    name = "global"
    config {
      envoy_prometheus_bind_addr = "0.0.0.0:1234"
    }
  }

  # Example support of other types (later)
  bootstrap {
    kind = "service-defaults"
    name = "web"
    protocol = "http"
    connect {
      enabled = true
    }
  }

  # Example support of non-bootstrap entries (later)
  file-managed {
    kind = "service-defaults"
    name = "api"
    protocol = "http"
    connect {
      enabled = true
    }
  }
}
```

With JSON translation:

```json
{
  "config_entries": {
    "bootstrap": [
      {
        "kind": "proxy_defaults",
        "name": "global",
        "config": {}
      },
      {
        "kind": "service_defaults",
        "name": "web",
        "protocol": true,
        "connect": {
          "enabled": true
        }
      }
    ],
    "file-managed": [
      {
        "kind": "service_defaults",
        "name": "api",
        "protocol": true,
        "connect": {
          "enabled": true
        }
      }
    ]
  }
}
```

@mkeeler does this make sense? I think this fits well with your CLI PR which effectively reads files in the same format as the nested format here. The benefit of `config_entries` block here goes even beyond the future possibility of non-bootstrap entries and would allow us to have something like:

```hcl
config_entries {
  from_dir = "/etc.d/consul/conf.d"
  # or
  bootstrap_from_dir = "/etc.d/consul/config_bootstrap"
}
```

**Comment by mkeeler** [comment]
@banks I think the proposed config changes look good. I still don’t exactly grasp why the map[interface{}]interface{} can exist with all the map walking being done. Regardless I think your idea for the config is better

**Comment by banks** [comment]
Closing this - I fixed the blocking issue in #5744 and Matt made #5746 to track the longer term non-blocking config issue.
